### ISSUE TITLE
`ShaderModule` requires a lot of manual initialization

### ISSUE THREAD

**Issue by Seelengrab** [JuliaGPU/Vulkan.jl#20]
E.g. I had to do `ShaderModule(device, length(code), reinterpret(UInt32, code))` instead of simply `ShaderModule(device, code)`, as I expected. 

I read the compiled shader code (done with `glslc`, like [vulkan-tutorial.com](https://vulkan-tutorial.com/Drawing_a_triangle/Graphics_pipeline_basics/Shader_modules) recommends) simply via `read("shader.frag")`, which returns a byte vector. Could maybe also be simplified to just make `ShaderModule(device, ::String)` take a path, though that would preclude hard-coded shaders in the same source code and dynamically generated shaders.

**Comment by Seelengrab** [comment]
From this point on, I think there will be quite a few "not yet implemented" sorts of issues - e.g. Pipeline creation also seems to require manual assembly of all `*Info` structs, as well as manual calling/unwrapping at `create_graphics_pipelines` level.

**Comment by Seelengrab** [comment]
I'm also not 100% sure how ShaderModule should be handled in concern with reference counting - if I destroy them explicitly at the end of my function (since the tutorial says that its fine for me to do so), Vulkan seems to yell at me when the app closes. This may go away once I add actual graphics pipeline code though, not 100% sure.

**Comment by serenity4** [comment]
There are some things that cannot be canonically wrapped, or that have corner cases make it very hard to be handled automatically.

Pipelines are created in batches, i.e. you get a vector of pipelines when unwrapped from `create_graphics_pipelines`, so we can't stick a constructor for `Pipeline`. Also, it takes an array of create info structures, and exposing arrays inputs for all arguments doesn't really make sense. It's better in that case to explicitly create the structures in the application side and just pass them in.

In places where we automatically unwrap elements (currently, with handle constructors), any error will just be thrown. In many situations it would be safer to actually manually unwrap, checking with `iserror` and unwrapping errors with `unwrap_error` (see the documentation for an example), and handle the cases where the result is not valid; particularly in performance-critical sections where the overhead of exceptions (try/catch blocks) is too high to be an option. That is why this `Result` type from ResultTypes.jl was introduced. Having to call the `create_...` constructors is not an issue at all. Having constructors already defined for e.g. `Instance` is nice but already at the limit of the scope of the package.

Concerning the finalization of `ShaderModule`, it is indeed not necessary to keep it alive once consumed by the API. As soon as it goes out of scope and the garbage collector runs, it will be automatically freed. If you finalized it yourself, then it will error, because it will be freed twice: once explicitly by you, and once when its refcount goes to zero.
The other way around is more tricky: when you have to preserve an object because a command implicitly depends on it (like when you submit work to a `Queue` with an `Image` being written to for example). In that case you have to `GC.@preserve` all the implicitly dependent variables during the call. There's no trivial solution for this.

> E.g. I had to do ShaderModule(device, length(code), reinterpret(UInt32, code)) instead of simply ShaderModule(device, code), as I expected.
I read the compiled shader code (done with glslc, like vulkan-tutorial.com recommends) simply via read("shader.frag"), which returns a byte vector. Could maybe also be simplified to just make ShaderModule(device, ::String) take a path, though that would preclude hard-coded shaders in the same source code and dynamically generated shaders.

There are a limited number places where you have to use this pattern. It could in theory be automated, but that's not in the scope of this package. It can be tempting to automate as much of these actions as possible, but it grows the complexity of the generator (which is already quite complex) and abstracts things that a Vulkan developer may not want to. This package is supposed to be a developer package, that helps you use the Vulkan API; it does not aim to wrap API commands beyond the basics. That should be done in other packages.

For shaders, I also wanted to make it very easy so that you take a shader text/io/path as input and produce a shader. That's what I did in the WIP package [VulkanShaders.jl](https://github.com/serenity4/VulkanShaders.jl).

So, to summarize, all these ideas are good and make total sense in higher-level packages, but not here where we only want to make it possible to work with Vulkan without the headaches from C-Julia interfacing.

**Comment by serenity4** [comment]
Also, keep in mind that you will have to write a lot of code for seemingly basic things. Vulkan is an explicit and verbose API, and is really huge. The possibilities it offers are almost endless. As soon as one tries to abstract things, it shrinks what you can do with it, although you may not notice it; because there will always be a lot of use cases that you didn't think about. Here we want to preserve all this power, and make it as accessible as we can, and we *really* want to avoid restricting functionality.

**Comment by Seelengrab** [comment]
Yep, I was thinking that I'm hitting the edges of what can be done without abstracting away too many things. I'm not familiar with how the generator works, so I can only naively suggest that maybe passing the length manually to `ShaderModule` is not intended, since we don't seem to have to anywhere else in the package.

Nevertheless, should I keep this open?

**Comment by serenity4** [comment]
Here, the length needs to be given in bytes, and the code in 4-bytes blobs. You may pass an array of `UInt64`/`UInt8`/`UInt32`, it will get converted to a `Ptr{UInt32}` pointer automatically. But since the array may come in any form, it is not currently possible to get the number of bytes, at least not without a significant change in infrastructure for the generator, which does not seem worth the trouble.

In case you are familiar with C interop, it's basically `cconvert`ing the pointer/array inputs, preserving them in a `deps` array, and `unsafe_convert`ing the required arguments. The [features](https://juliagpu.github.io/Vulkan.jl/dev/features/#Structures) section of the docs has more details if you want. To support the above we'd need to add more steps in the body of the function, and interact with intermediate variables, which quickly becomes hairy.

Unless there is something else you'd like to discuss related to this issue, feel free to close it.

**Comment by Seelengrab** [comment]
Ooh, I see! I didn't even try, but it works just as well without `reinterpret(UInt32, code)`, yes. I think I'm going to keep it in there though, for clarities sake.
